// @snapshot-labs/snapshot.js/src/strategies/mushrooms/dist/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
exports.__esModule = true;
exports.strategy = exports.examples = exports.version = exports.author = void 0;
var units_1 = require("@ethersproject/units");
var multicaller_1 = require("../../utils/multicaller");
var examples_json_1 = require("./examples.json");
exports.author = '@MushroomsFinan1';
exports.version = '0.1.0';
exports.examples = examples_json_1["default"];
var erc20Abi = [
    {
        inputs: [],
        name: 'totalSupply',
        outputs: [
            {
                internalType: 'uint256',
                name: '',
                type: 'uint256'
            }
        ],
        stateMutability: 'view',
        type: 'function'
    },
    {
        inputs: [
            {
                internalType: 'address',
                name: 'account',
                type: 'address'
            }
        ],
        name: 'balanceOf',
        outputs: [
            {
                internalType: 'uint256',
                name: '',
                type: 'uint256'
            }
        ],
        stateMutability: 'view',
        type: 'function'
    }
];
var masterChefAbi = [
    {
        inputs: [
            {
                internalType: 'uint256',
                name: '',
                type: 'uint256'
            },
            {
                internalType: 'address',
                name: '',
                type: 'address'
            }
        ],
        name: 'userInfo',
        outputs: [
            {
                internalType: 'uint256',
                name: 'amount',
                type: 'uint256'
            },
            {
                internalType: 'uint256',
                name: 'rewardDebt',
                type: 'uint256'
            }
        ],
        stateMutability: 'view',
        type: 'function'
    },
    {
        inputs: [
            {
                internalType: 'uint256',
                name: '',
                type: 'uint256'
            }
        ],
        name: 'poolInfo',
        outputs: [
            {
                internalType: 'contract IERC20',
                name: 'lpToken',
                type: 'address'
            },
            {
                internalType: 'uint256',
                name: 'allocPoint',
                type: 'uint256'
            },
            {
                internalType: 'uint256',
                name: 'lastRewardBlock',
                type: 'uint256'
            },
            {
                internalType: 'uint256',
                name: 'accMMPerShare',
                type: 'uint256'
            },
            {
                internalType: 'uint256',
                name: 'amount',
                type: 'uint256'
            }
        ],
        stateMutability: 'view',
        type: 'function'
    }
];
function strategy(space, network, provider, addresses, options, snapshot) {
    return __awaiter(this, void 0, void 0, function () {
        var blockTag, masterChefMulti, masterChefResult_1, erc20Multi, erc20Result_1, masterChefResult_2;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    blockTag = typeof snapshot === 'number' ? snapshot : 'latest';
                    masterChefMulti = new multicaller_1["default"](network, provider, masterChefAbi, {
                        blockTag: blockTag
                    });
                    addresses.forEach(function (address) {
                        masterChefMulti.call(address + ".userInfo.amount", options.masterchef, 'userInfo', [options.pool, address]);
                    });
                    if (!(options.type === 'lp')) return [3 /*break*/, 3];
                    masterChefMulti.call('poolInfo.lpToken', options.masterchef, 'poolInfo', [
                        options.pool
                    ]);
                    return [4 /*yield*/, masterChefMulti.execute()];
                case 1:
                    masterChefResult_1 = _a.sent();
                    erc20Multi = new multicaller_1["default"](network, provider, erc20Abi, {
                        blockTag: blockTag
                    });
                    erc20Multi.call('lpTotalSupply', masterChefResult_1.poolInfo.lpToken[0], 'totalSupply');
                    erc20Multi.call('poolMMBalance', options.govtoken, 'balanceOf', [
                        masterChefResult_1.poolInfo.lpToken[0]
                    ]);
                    return [4 /*yield*/, erc20Multi.execute()];
                case 2:
                    erc20Result_1 = _a.sent();
                    return [2 /*return*/, Object.fromEntries(addresses.map(function (address) {
                            return [
                                address,
                                parseFloat(units_1.formatUnits(masterChefResult_1[address].userInfo.amount[0]
                                    .mul(erc20Result_1.poolMMBalance)
                                    .div(erc20Result_1.lpTotalSupply)
                                    .toString(), 18))
                            ];
                        }))];
                case 3: return [4 /*yield*/, masterChefMulti.execute()];
                case 4:
                    masterChefResult_2 = _a.sent();
                    return [2 /*return*/, Object.fromEntries(addresses.map(function (address) {
                            return [
                                address,
                                parseFloat(units_1.formatUnits(masterChefResult_2[address].userInfo.amount.toString(), 18))
                            ];
                        }))];
            }
        });
    });
}
exports.strategy = strategy;
