// @snapshot-labs/snapshot.js/src/strategies/flashstake/dist/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
exports.__esModule = true;
exports.strategy = exports.examples = exports.version = exports.author = void 0;
var address_1 = require("@ethersproject/address");
var utils_1 = require("../../utils");
var examples_json_1 = require("./examples.json");
var FLASHSTAKE_SUBGRAPH_URL = {
    '1': 'https://api.thegraph.com/subgraphs/name/blockzerohello/flash-stake-stats-v2-subgraph'
};
exports.author = 'anassohail99';
exports.version = '0.1.0';
exports.examples = examples_json_1["default"];
function strategy(_space, network, _provider, addresses, options, snapshot) {
    return __awaiter(this, void 0, void 0, function () {
        var params, params2, result, stakesResult, score;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    params = {
                        users: {
                            __args: {
                                where: {
                                    id_in: addresses.map(function (address) { return address.toLowerCase(); })
                                },
                                first: 1000
                            },
                            id: true,
                            liquidityPositions: {
                                liquidityTokenBalance: true,
                                user: {
                                    id: true
                                },
                                pair: {
                                    id: true,
                                    token0: {
                                        id: true
                                    },
                                    reserve0: true,
                                    token1: {
                                        id: true
                                    },
                                    reserve1: true,
                                    totalSupply: true
                                }
                            }
                        }
                    };
                    params2 = {
                        stakes: {
                            __args: {
                                where: {
                                    user_in: addresses.map(function (address) { return address.toLowerCase(); }),
                                    isActive: true
                                }
                            },
                            id: true,
                            amountIn: true,
                            user: true
                        }
                    };
                    if (snapshot !== 'latest') {
                        // @ts-ignore
                        params.users.__args.block = { number: snapshot };
                        // @ts-ignore
                        params2.stakes.__args.block = { number: snapshot };
                    }
                    return [4 /*yield*/, utils_1.subgraphRequest(FLASHSTAKE_SUBGRAPH_URL[network], params)];
                case 1:
                    result = _a.sent();
                    return [4 /*yield*/, utils_1.subgraphRequest(FLASHSTAKE_SUBGRAPH_URL[network], params2)];
                case 2:
                    stakesResult = _a.sent();
                    score = {};
                    if (stakesResult && stakesResult.stakes) {
                        stakesResult.stakes.map(function (_data) {
                            var address = address_1.getAddress(_data.user);
                            if (!score[address])
                                score[address] = 0;
                            score[address] = Number(score[address]) + Number(_data.amountIn);
                        });
                    }
                    if (result && result.users) {
                        result.users.map(function (_data) {
                            var _a;
                            if ((_a = _data.liquidityPositions[0]) === null || _a === void 0 ? void 0 : _a.pair) {
                                _data.liquidityPositions.map(function (__data) {
                                    var address = address_1.getAddress(__data.user.id);
                                    var token0perFlash = Number(__data.pair.reserve0) / Number(__data.pair.totalSupply);
                                    var userScore = token0perFlash * Number(__data.liquidityTokenBalance);
                                    if (!score[address])
                                        score[address] = 0;
                                    score[address] = Number(score[address]) + userScore;
                                });
                            }
                        });
                    }
                    return [2 /*return*/, score || {}];
            }
        });
    });
}
exports.strategy = strategy;
