// @snapshot-labs/snapshot.js/src/strategies/spookyswap/dist/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
exports.__esModule = true;
exports.strategy = exports.examples = exports.version = exports.author = void 0;
var units_1 = require("@ethersproject/units");
var multicaller_1 = require("../../utils/multicaller");
var examples_json_1 = require("./examples.json");
exports.author = 'ooGwei';
exports.version = '0.1.0';
exports.examples = examples_json_1["default"];
var FARM_ADDRESS = '0x2b2929E785374c651a81A63878Ab22742656DcDd';
var LP_TOKEN_ADDRESS = '0xEc7178F4C41f346b2721907F5cF7628E388A7a58';
var BOO_TOKEN_ADDRESS = '0x841FAD6EAe12c286d1Fd18d1d525DFfA75C7EFFE';
var abi = [
    {
        inputs: [
            {
                internalType: 'address',
                name: '',
                type: 'address'
            }
        ],
        name: 'balanceOf',
        outputs: [
            {
                internalType: 'uint256',
                name: 'amount',
                type: 'uint256'
            }
        ],
        stateMutability: 'view',
        type: 'function'
    },
    {
        inputs: [
            {
                internalType: 'uint256',
                name: '',
                type: 'uint256'
            },
            {
                internalType: 'address',
                name: '',
                type: 'address'
            }
        ],
        name: 'userInfo',
        outputs: [
            {
                internalType: 'uint256',
                name: 'amount',
                type: 'uint256'
            },
            {
                internalType: 'uint256',
                name: 'rewardDebt',
                type: 'uint256'
            }
        ],
        stateMutability: 'view',
        type: 'function'
    },
    {
        constant: true,
        inputs: [],
        name: 'totalSupply',
        outputs: [
            {
                internalType: 'uint256',
                name: '',
                type: 'uint256'
            }
        ],
        payable: false,
        stateMutability: 'view',
        type: 'function'
    }
];
function strategy(space, network, provider, addresses, options, snapshot) {
    return __awaiter(this, void 0, void 0, function () {
        var blockTag, multi, result;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    blockTag = typeof snapshot === 'number' ? snapshot : 'latest';
                    multi = new multicaller_1["default"](network, provider, abi, { blockTag: blockTag });
                    addresses.forEach(function (address) {
                        multi.call("boo." + address, BOO_TOKEN_ADDRESS, 'balanceOf', [address]);
                        multi.call("lpInFarm." + address, FARM_ADDRESS, 'userInfo', ['0', address]);
                        multi.call("lp." + address, LP_TOKEN_ADDRESS, 'balanceOf', [address]);
                        options.vaultTokens.forEach(function (token) {
                            multi.call("vaultTokens." + address + "." + token.address, token.address, 'balanceOf', [address]);
                        });
                    });
                    multi.call("lp.totalSupply", LP_TOKEN_ADDRESS, 'totalSupply', []);
                    multi.call("lp.boo", BOO_TOKEN_ADDRESS, 'balanceOf', [LP_TOKEN_ADDRESS]);
                    return [4 /*yield*/, multi.execute()];
                case 1:
                    result = _a.sent();
                    return [2 /*return*/, Object.fromEntries(addresses.map(function (address) { return [
                            address,
                            parseFloat(units_1.formatUnits(result.boo[address]
                                .mul(options.boo.numerator)
                                .div(options.boo.denominator), 18)) +
                                parseFloat(units_1.formatUnits(result.lpInFarm[address][0]
                                    .mul(result.lp.boo)
                                    .div(result.lp.totalSupply)
                                    .mul(options.lp.numerator)
                                    .div(options.lp.denominator), 18)) +
                                parseFloat(units_1.formatUnits(result.lp[address]
                                    .mul(result.lp.boo)
                                    .div(result.lp.totalSupply)
                                    .mul(options.lp.numerator)
                                    .div(options.lp.denominator), 18)) +
                                options.vaultTokens.reduce(function (prev, token, idx) {
                                    return prev +
                                        parseFloat(units_1.formatUnits(result.vaultTokens[address][token.address]
                                            .mul(options.vaultTokens[idx].numerator)
                                            .div(options.vaultTokens[idx].denominator), options.vaultTokens[idx].decimal));
                                }, 0)
                        ]; }))];
            }
        });
    });
}
exports.strategy = strategy;
